vowpalwabbit/parse_args.cc

```diff
@@ -529,6 +529,12 @@ void parse_feature_tweaks(
 
   option_group_definition feature_options("Feature options");
   feature_options
+      .add(make_option("privacy_activation", all.privacy_activation)
```

Comment:
It looks like this isn't supported in every base learner. However, for a feature such as this I think it's very important to fail if a user requested `privacy_activation` but the stack's base learner wont do it

Reply:
We can check again but last time we checked all base learners would eventually use the gd methods that have been altered here and the tag is setup by setup base so it looks like all base learners can support the feature

Reply:
`gd_mf` is one which will be unsupported. Also, how does this interact with `print` and `sender`?

Reply:
`gd_mf` and `print` use `GD::foreach_feature` as far as I can see so this change should be supported there. I am not sure what `sender` does but it doesn't look like it is using weights at all.

We are looking into an alternative implementation with @manavsinghal157 of putting this change into the weight's `[]` operator so then any reduction using the `weights` will get this feature for sure (need to check the perf on that change)

Do you think it will still be a problem in that case? I am worried that if we have to add a warning in all the base learners then we might add a new base learner and forget to add the warning there so it's best to make sure this feature is used in all base learners

Reply:
that warning system could be generic and done once for all args, it would just be maybe out of scope for this PR?

We would have to keep track of args that were parsed but their 'root' necessary arg was not fulfilled. That would imply it's an error, and right now i think we quietly fail to report that the arg didnt have a side-effect on the instantiation. It gets tricky since some args are repeated across reductions, but shouldn't be that complicated.

Reply:
You're right, I missed that in `gd_mf`! VW has a bit of a problem with knowing if an argument you supplied is actually used or not, hence my concern with an option centered around ensuring privacy. But really, this issue is not created by or limited to this new option. Your explanation is good, it seems we are sufficiently covered here. In fact this is probably one of the more universal options given the nature of it being in the base of the stack.

Hmm, `sender` is used with daemon mode so it isn't really a base learner in the traditional sense.

Reply:
This is the alternative approach [here](https://github.com/VowpalWabbit/vowpal_wabbit/pull/3334/files#diff-f75d24249a7ddfb0203db0f25db4adec31f15f48abcf32bed9fef843f9496bfcR170)